torch/quasirandom.py

```diff
@@ -0,0 +1,111 @@
+    a maximum dimension of 1111. It uses direction numbers to generate these sequences, and these
+    numbers have been adapted from `here <http://web.maths.unsw.edu.au/~fkuo/sobol/joe-kuo-old.1111>`_.
+
+    References:
+      - Art B. Owen. Scrambling Sobol and Niederreiter-Xing points. Journal of
+        Complexity, 14(4):466-489, December 1998.
+
+      - I. M. Sobol. The distribution of points in a cube and the accurate
+        evaluation of integrals. Zh. Vychisl. Mat. i Mat. Phys., 7:784-802, 1967.
+
+    Args:
+        dimension (Int): The dimensionality of the sequence to be drawn
+        scramble (bool, optional): Setting this to ``True`` will produce scrambled Sobol sequences.
+                                   Scrambling is capable of producing better Sobol sequences.
+                                   Default: ``False``.
+        seed (Int, optional): This is the seed for the scrambling. The seed of the random number
+                              generator is set to this, if specified. Default: ``None``
+        device: Device string. This sets the saves the states of the ``SobolEngine`` in :attr:`device`.
+                Subsequent sampling / fast-forwarding / resetting is done on :attr:`device`. Default: 'cpu'
+
+    Examples::
+
+        >>> soboleng = torch.quasirandom.SobolEngine(dimension=5, scramble=False)
+        >>> soboleng.draw(3)
+        tensor([[0.5000, 0.5000, 0.5000, 0.5000, 0.5000],
+                [0.7500, 0.2500, 0.7500, 0.2500, 0.7500],
+                [0.2500, 0.7500, 0.2500, 0.7500, 0.2500]])
+    """
+    MAXBIT = 30
+
+    def __init__(self, dimension, scramble=False, seed=None, device='cpu'):
+        if dimension > 1111 or dimension < 1:
+            raise ValueError("Supported range of dimensionality for SobolEngine is [1, 1111]")
+
+        self.seed = seed
+        self.scramble = scramble
+        self.dimension = dimension
+        self.device = device
+
+        self.sobolstate = torch.zeros(dimension, self.MAXBIT, dtype=torch.long, device=self.device)
+        torch._sobol_engine_initialize_state_(self.sobolstate, self.dimension)
+
+        if self.scramble:
+            g = torch.Generator()
+            if self.seed is not None:
+                g.manual_seed(self.seed)
+
+            self.shift = torch.mv(torch.randint(2, (self.dimension, self.MAXBIT), device=self.device, generator=g),
+                                  torch.pow(2, torch.arange(0, self.MAXBIT, device=self.device)))
+
+            ltm = torch.randint(2, (self.dimension, self.MAXBIT, self.MAXBIT),
+                                generator=g, device=self.device).tril()
+
+            torch._sobol_engine_scramble_(self.sobolstate, ltm, self.dimension)
+        else:
+            self.shift = torch.zeros(self.dimension, dtype=torch.long, device=self.device)
+
+        self.quasi = self.shift.clone()
+        self.num_generated = 0
+
+    def draw(self, n=1):
```

Comment:
Could we extend this interface so that
1. it takes an optional `out` tensor that is filled with the result (rather than returning a new one). This mimics the interface of `torch.rand` and many other operations.
2. it takes a `dtype` arg, and returns the samples of in that data type.

This should be pretty straightforward to do and would make this interface more convenient to work with.